hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/DefaultCompactWriter.java

```diff
@@ -0,0 +1,481 @@
+import static com.hazelcast.nio.serialization.FieldType.TIME;
+import static com.hazelcast.nio.serialization.FieldType.TIMESTAMP;
+import static com.hazelcast.nio.serialization.FieldType.TIMESTAMP_ARRAY;
+import static com.hazelcast.nio.serialization.FieldType.TIMESTAMP_WITH_TIMEZONE;
+import static com.hazelcast.nio.serialization.FieldType.TIMESTAMP_WITH_TIMEZONE_ARRAY;
+import static com.hazelcast.nio.serialization.FieldType.TIME_ARRAY;
+import static com.hazelcast.nio.serialization.FieldType.UTF;
+import static com.hazelcast.nio.serialization.FieldType.UTF_ARRAY;
+
+public class DefaultCompactWriter implements CompactWriter {
+
+    private final CompactStreamSerializer serializer;
+    private final Schema schema;
+    private final BufferObjectDataOutput out;
+    private final int offset;
+    private final int[] fieldPositions;
+    private final boolean includeSchemaOnBinary;
+
+    public DefaultCompactWriter(CompactStreamSerializer serializer,
+                                BufferObjectDataOutput out, Schema schema, boolean includeSchemaOnBinary) {
+        this.serializer = serializer;
+        this.out = out;
+        this.schema = schema;
+        if (schema.getNumberOfVariableLengthFields() != 0) {
+            this.fieldPositions = new int[schema.getNumberOfVariableLengthFields()];
+            offset = out.position() + INT_SIZE_IN_BYTES;
+            //skip for length and primitives
+            out.writeZeroBytes(schema.getPrimitivesLength() + INT_SIZE_IN_BYTES);
+        } else {
+            this.fieldPositions = null;
+            offset = out.position();
+            //skip for  primitives
+            out.writeZeroBytes(schema.getPrimitivesLength());
+        }
+        this.includeSchemaOnBinary = includeSchemaOnBinary;
+    }
+
+    public byte[] toByteArray() {
+        return out.toByteArray();
+    }
+
+    public void end() {
+        try {
+            if (schema.getNumberOfVariableLengthFields() == 0) {
+                //There are no var length
+                return;
+            }
+            for (int i = fieldPositions.length - 1; i >= 0; i--) {
+                out.writeInt(fieldPositions[i]);
+            }
+            int position = out.position();
+            int length = position - offset;
+            //write length
+            out.writeInt(offset - INT_SIZE_IN_BYTES, length);
+        } catch (IOException e) {
+            throw illegalStateException(e);
+        }
+    }
+
+    @Override
+    public void writeInt(@Nonnull String fieldName, int value) {
+        int position = getFixedLengthFieldPosition(fieldName, INT);
+        try {
+            out.writeInt(position, value);
+        } catch (IOException e) {
+            throw illegalStateException(e);
+        }
+    }
+
+    @Override
+    public void writeLong(@Nonnull String fieldName, long value) {
+        int position = getFixedLengthFieldPosition(fieldName, LONG);
+        try {
+            out.writeLong(position, value);
```

Comment:
Will we have the compact serialization force use only the LITTLE_ENDIAN? The output stream here is BIG_ENDIAN by default. We need to handle these.

Reply:
I am resolving the conversation. We have decided to go with LITTLE_ENDIAN for the new format.